lib/core/src/dirty.rs

```diff
@@ -1,110 +1,544 @@
+
+// FIXME: This is very error prone. Fix it after this gets resolved:
+// https://github.com/rust-lang/rust/issues/66505
+
+// [1] Please refer to `Prelude::drop_lifetime` docs to learn why it is safe to
+// use it here.
+impl<T, OnSet> SharedDirtyFlag<T,OnSet>
+where for<'t> &'t T: IntoIterator {
+    pub fn iter(&self) -> SharedDirtyFlagIter<T, OnSet> {
+        let _borrow   = self.rc.borrow();
+        let reference = unsafe { drop_lifetime(&_borrow) }; // [1]
+        let iter      = reference.into_iter();
+        SharedDirtyFlagIter { iter, _borrow }
+    }
+}
+
+/// Iterator guard for SharedDirtyFlag. It exposes the iterator of original
+/// structure behind the shared reference.
+pub struct SharedDirtyFlagIter<'t,T,OnSet>
+where &'t T: IntoIterator {
+    pub iter : <&'t T as IntoIterator>::IntoIter,
+    _borrow  : Ref<'t,DirtyFlag<T,OnSet>>
+}
+
+impl<'t,T,OnSet> Deref for SharedDirtyFlagIter<'t,T,OnSet>
+where &'t T: IntoIterator {
+    type Target = <&'t T as IntoIterator>::IntoIter;
+    fn deref(&self) -> &Self::Target { &self.iter }
+}
+
+impl<'t,T,OnSet> DerefMut for SharedDirtyFlagIter<'t,T,OnSet>
+where &'t T: IntoIterator {
+    fn deref_mut(&mut self) -> &mut Self::Target { &mut self.iter }
+}
+
+impl<'t,T,OnSet> Iterator for SharedDirtyFlagIter<'t,T,OnSet>
+where &'t T: IntoIterator {
+    type Item = <&'t T as IntoIterator>::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
+
+// =============================================================================
+// === Flags ===================================================================
+// =============================================================================
+
+// ============
+// === Bool ===
+// ============
+
+/// The on / off dirty flag. If you need a simple dirty / clean switch, this one
+/// is the right choice.
+
+pub type  Bool       <OnSet> = DirtyFlag       <BoolData,OnSet>;
+pub type  SharedBool <OnSet> = SharedDirtyFlag <BoolData,OnSet>;
+pub trait BoolCtx    <OnSet> = where OnSet: Callback0;
+
+#[derive(Debug,Display,Default)]
+pub struct BoolData { is_dirty: bool }
+impl DirtyFlagData for BoolData {
+    type SetArgs = ();
+    fn raw_check     (&self) -> bool        { self.is_dirty }
+    fn raw_check_for (&self, _:&()) -> bool { self.is_dirty }
+    fn raw_set       (&mut self, _:())      { self.is_dirty = true  }
+    fn raw_unset     (&mut self)            { self.is_dirty = false }
+}
+
+
 // =============
-// === Dirty ===
+// === Range ===
 // =============
 
-#[derive(Clone, Debug)]
-pub struct Dirty {
-    pub data: Rc<RefCell<DirtyData>>,
+/// Dirty flag which keeps information about a range of dirty items. It does not
+/// track items separately, nor you are allowed to keep multiple ranges in it.
+/// Just a single value range.
+
+pub type  Range       <Ix,OnSet> = DirtyFlag       <RangeData<Ix>,OnSet>;
```

Comment:
Generic parameters are some kind of the same concept for each type :P Such alignment is common for FP langs, like Haskell. It's easy to see now that several types share a type parameter that is responsible for the same thing. I think it may be a personal taste here, but this discussion shows me that we need to create a formal document about code structure sooner than later (the current one assumes we are using auto-formatter, which we are not). We should also have sometimes calls about situations where we do not agree to choose a common strategy then.